File: java/src/main/java/org/mydotey/objectpool/threadpool/DefaultThreadPool.java
package org.mydotey.objectpool.threadpool;

import java.io.IOException;
import java.util.Objects;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

import org.mydotey.objectpool.ObjectPool;
import org.mydotey.objectpool.ObjectPoolConfig;
import org.mydotey.objectpool.ObjectPool.Entry;
import org.mydotey.objectpool.facade.ObjectPools;

/**
 * @author koqizhao
 *
 * Feb 6, 2018
 */
public class DefaultThreadPool implements ThreadPool {

    private ThreadPoolConfig _config;
    protected ObjectPool<WorkerThread> _objectPool;

    protected boolean _hasQueue;
    protected BlockingQueue<Runnable> _taskQueue;
    protected Thread _taskConsumer;

    public DefaultThreadPool(ThreadPoolConfig config) {
        Objects.requireNonNull(config, "config is null");
        _config = config;
        _objectPool = newObjectPool();

        _hasQueue = _config.getQueueCapacity() > 0;
        if (_hasQueue) {
            _taskQueue = new LinkedBlockingQueue<>(_config.getQueueCapacity());
            _taskConsumer = new Thread(() -> consumeTask());
            _taskConsumer.setDaemon(true);
            _taskConsumer.start();
        }
    }

    protected ObjectPool<WorkerThread> newObjectPool() {
        ObjectPoolConfig.Builder<WorkerThread> builder = ObjectPools.newObjectPoolConfigBuilder();
        setObjectPoolConfigBuilder(builder);
        return ObjectPools.newObjectPool(builder.build());
    }

    protected void setObjectPoolConfigBuilder(ObjectPoolConfig.AbstractBuilder<WorkerThread, ?> builder) {
        builder.setMaxSize(_config.getMaxSize()).setMinSize(_config.getMinSize())
                .setObjectFactory(() -> new WorkerThread(t -> getObjectPool().release(t.getPoolEntry())))
                .setOnCreate(e -> {
                    e.getObject().setPoolEntry(e);
                    e.getObject().start();
                }).setOnClose(e -> e.getObject().interrupt());
    }

    protected ObjectPool<WorkerThread> getObjectPool() {
        return _objectPool;
    }

    @Override
    public ThreadPoolConfig getConfig() {
        return _config;
    }

    @Override
    public int getSize() {
        return getObjectPool().getSize();
    }

    @Override
    public int getQueueSize() {
        return _hasQueue ? _taskQueue.size() : 0;
    }

    @Override
    public void submit(Runnable task) throws InterruptedException {
        Objects.requireNonNull(task, "task is null");

        if (_hasQueue) {
            if (trySubmit(task))
                return;

            _taskQueue.put(task);
            return;
        }

        Entry<WorkerThread> entry = getObjectPool().acquire();
        entry.getObject().setTask(task);
    }

    @Override
    public boolean trySubmit(Runnable task) {
        Objects.requireNonNull(task, "task is null");

        Entry<WorkerThread> entry = getObjectPool().tryAcquire();
        if (entry == null)
            return false;

        entry.getObject().setTask(task);
        return true;
    }

    protected void consumeTask() {
        while (!Thread.currentThread().isInterrupted()) {
            try {
                Runnable task = _taskQueue.take();
                Entry<WorkerThread> entry = getObjectPool().acquire();
                entry.getObject().setTask(task);
            } catch (Exception e) {
                break;
            }
        }
    }

    @Override
    public void close() throws IOException {
        getObjectPool().close();

        if (_hasQueue)
            _taskConsumer.interrupt();
    }
}

File: dotnet/src/MyDotey.ObjectPool/ThreadPool/DefaultThreadPool.cs
using System;
using System.Collections.Concurrent;
using System.Threading;
using MyDotey.ObjectPool.Facade;

/**
 * @author koqizhao
 *
 * Feb 22, 2018
 */
namespace MyDotey.ObjectPool.ThreadPool
{
    public class DefaultThreadPool : IThreadPool
    {
        public virtual IThreadPoolConfig Config { get; }
        protected internal virtual IObjectPool<WorkerThread> ObjectPool { get; }

        protected virtual bool HasQueue { get; }
        protected virtual BlockingCollection<Action> TaskQueue { get; }
        protected virtual Thread TaskConsumer { get; }

        public DefaultThreadPool(IThreadPoolConfig config)
        {
            if (config == null)
                throw new ArgumentNullException("config is null");

            Config = config;
            ObjectPool = NewObjectPool();

            HasQueue = Config.QueueCapacity > 0;
            if (HasQueue)
            {
                TaskQueue = new BlockingCollection<Action>(new ConcurrentQueue<Action>(), Config.QueueCapacity);
                TaskConsumer = new Thread(ConsumeTask)
                {
                    IsBackground = true
                };
                TaskConsumer.Start();
            }
        }

        protected virtual IObjectPool<WorkerThread> NewObjectPool()
        {
            IBuilder<WorkerThread> builder = ObjectPools.NewObjectPoolConfigBuilder<WorkerThread>();
            SetObjectPoolConfigBuilder<IBuilder<WorkerThread>>(builder);
            return ObjectPools.NewObjectPool(builder.Build());
        }

        protected virtual void SetObjectPoolConfigBuilder<B>(IAbstractBuilder<WorkerThread, B> builder)
            where B : IAbstractBuilder<WorkerThread, B>
        {
            builder.SetMaxSize(Config.MaxSize)
                    .SetMinSize(Config.MinSize)
                    .SetObjectFactory(() => new WorkerThread(t => ObjectPool.Release(t.PoolEntry)))
                    .SetOnCreate(e =>
                    {
                        e.Object.PoolEntry = e;
                        e.Object.Start();
                    })
                    .SetOnClose(e => e.Object.InnerThread.Interrupt());
        }

        public virtual int Size { get { return ObjectPool.Size; } }

        public virtual int QueueSize { get { return HasQueue ? TaskQueue.Count : 0; } }

        public virtual void Submit(Action task)
        {
            if (task == null)
                throw new ArgumentNullException("task is null");

            if (HasQueue)
            {
                if (TrySubmit(task))
                    return;

                TaskQueue.Add(task);
                return;
            }

            IEntry<WorkerThread> entry = ObjectPool.Acquire();
            entry.Object.SetTask(task);
        }

        public virtual bool TrySubmit(Action task)
        {
            if (task == null)
                throw new ArgumentNullException("task is null");

            IEntry<WorkerThread> entry = ObjectPool.TryAcquire();
            if (entry == null)
                return false;

            entry.Object.SetTask(task);
            return true;
        }

        protected virtual void ConsumeTask()
        {
            while (true)
            {
                try
                {
                    Action task = TaskQueue.Take();
                    IEntry<WorkerThread> entry = ObjectPool.Acquire();
                    entry.Object.SetTask(task);
                }
                catch (Exception ex)
                {
                    //
                    break;
                }
            }
        }

        public virtual void Dispose()
        {
            ObjectPool.Dispose();

            if (HasQueue)
                TaskConsumer.Interrupt();
        }
    }
}